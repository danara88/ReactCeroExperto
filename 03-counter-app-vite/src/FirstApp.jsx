// Los propTypes nos ayuda a ponerle tipo a las properties
import PropTypes from 'prop-types';

/**
 * NOTA:
 * Si no hay alguna dependencia de el componente y de alguna variable,
 * poner las variables y funciones fuera del componente.
 */

/**
 * Podemos pasar variables al HTML mientras no sean objetos
 */
const newMessage = {
  message: 'Hello',
  title: 'Daniel',
};

/**
 * Si la function getName lo pongo dentro de FirstApp,
 * al momento de redibujar el componente, va a volver a especificar
 * en memoria el espacio de la funcion.
 */
const getName = (value) => value;

const FirstApp = ({ title, subTitle }) => {
  // if (!title) {
  //     throw new Error("Title no existe");
  // }

  return (
    <>
      <h1 data-testid="test-title">{title}</h1>

      <p>{subTitle}</p>
      <p>{subTitle}</p>
      <code>{JSON.stringify(newMessage)}</code>
      <div>First App</div>
      <div>{getName('Daniel Aranda')}</div>
      {/* <p>soy un subtitulo</p> */}
    </>
  );
};

FirstApp.propTypes = {
  title: PropTypes.string.isRequired,
  subTitle: PropTypes.string.isRequired,
};

// FirstApp.defaultProps = {
//   title: 'No hay titulo',
//   subTitle: 0,
// };

export default FirstApp;
