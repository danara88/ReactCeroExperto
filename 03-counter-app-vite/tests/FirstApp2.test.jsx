import { describe, expect, test } from '@jest/globals';
import { render, screen } from '@testing-library/react';
import FirstApp from '../src/FirstApp';

/**
 * La prueba del snapshot se hace hasta que sabemos que el componente no recibira modificaciones.
 * Si el componente esta en desarrollo, no hacer la prueba del snapshot.
 *
 * screen: Es el objeto que estamos renderizando. Obtenemos el screen actualizado despues de llamar el 'render'
 */
describe('Pruebasd en <FirstApp />', () => {
  const title = 'Este es un titulo';
  const subTitle = 'Esto es un subtitulo';

  test('debe de hacer match con el snapshot', () => {
    // Arrange
    const { container } = render(<FirstApp title={title} subTitle={subTitle} />);
    // Act & Assert
    expect(container).toMatchSnapshot();
  });

  test('debe de mostrar el mensaje "Esto es un titulo"', () => {
    // Arrange
    render(<FirstApp title={title} subTitle={subTitle} />);
    // Act & Assert
    expect(screen.getByText(title)).toBeTruthy();
  });

  test('debe de mostrar el titulo en un tag h1', () => {
    // Arrange
    render(<FirstApp title={title} subTitle={subTitle} />);
    // Act & Assert
    expect(screen.getByRole('heading', { level: 1 }).innerHTML).toContain(title);
  });

  test('debe de mostrar el subtitulo enviado por props', () => {
    // Arrange
    render(<FirstApp title={title} subTitle={subTitle} />);
    // Act
    const result = screen.getAllByText(subTitle);
    // Assert
    expect(result.length).toBe(2);
  });
});
