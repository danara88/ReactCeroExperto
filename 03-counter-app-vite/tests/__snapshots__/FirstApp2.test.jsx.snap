// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebasd en <FirstApp /> debe de hacer match con el snapshot 1`] = `
<div>
  <h1
    data-testid="test-title"
  >
    Este es un titulo
  </h1>
  <p>
    Esto es un subtitulo
  </p>
  <p>
    Esto es un subtitulo
  </p>
  <code>
    {"message":"Hello","title":"Daniel"}
  </code>
  <div>
    First App
  </div>
  <div>
    Daniel Aranda
  </div>
</div>
`;
